my_2d_robot/src/joint_states_subscriber.cpp

```diff
@@ -1,19 +1,150 @@
+#include <string>
+#include <fstream>
+#include <sstream>
+#include <boost/filesystem.hpp>
+#include "yaml-cpp/yaml.h"
+#include <eigen3/Eigen/Dense>
 
-void joint_statesCallback(const std_msgs::String::ConstPtr& msg)
+using namespace Eigen;
+typedef Matrix<float, 2, 1> Vector2f;
+
+class SensorMeasurementData
+{
+public:
+  // Contructor.
+  SensorMeasurementData(ros::NodeHandle* n)
+  {
+    // Initialises length of links (specified in urdf).
+    link_1_ = 5.0;
+    link_2_ = 5.0;
+
+    sub_ = n->subscribe("joint_states", 1000, &SensorMeasurementData::joint_statesCallback, this);
+
+    // Gets the joint angle offsets which are set in the launch file.
+    n->getParam("Theta1_offset", theta1_offset_);
+    n->getParam("Theta2_offset", theta2_offset_);
+  }
+
+  // Deconstructor.
+  ~SensorMeasurementData()
+  {
+    std::stringstream stream;
+
+    std::ofstream fout;
+    fout.open(my_output_, std::fstream::app);
+
+    if (!fout)
+    {
+      ROS_ERROR("error");
+    }
+
+    stream << output_data_ << std::endl;
+    fout << stream.rdbuf();
+    fout.close();
+  }
+
+private:
+  ros::Subscriber sub_;
+  // Joint positions.
+  float position_joint1_;
+  float position_joint2_;
+  // Joint angle offsets.
+  float theta1_offset_;
+  float theta2_offset_;
+  // Link lengths.
+  float link_1_;
+  float link_2_;
+  // Vector to store the end effector position.
+  Vector2f ee_position_;
+  // String used to store package filepath.
+  std::string my_output_;
+  // String used to store parsed data from yaml-cpp library.
+  std::string output_data_ = "";
+
+  // Callback method to get joint angles.
+  void joint_statesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+
+  // Method to write the joint state angles and end effector position to yaml file.
+  void saveYaml();
+
+  // Method to calculate the position of the end effector.
+  Vector2f eePos(float joint_1, float joint_2)
+  {
+    // Includes the offsets for joint angles.
+    float Theta1 = joint_1 - theta1_offset_;
+    float Theta2 = joint_2 - theta2_offset_;
+
+    // Calculates the end effector position.
+    Vector2f position;
+    position << (link_1_ * cos(Theta1 * M_PI / 180)) + (link_2_ * cos((Theta1 + Theta2) * M_PI / 180)),
+        (link_1_ * sin(Theta1 * M_PI / 180)) + (link_2_ * sin((Theta1 + Theta2) * M_PI / 180));
+
+    return position;
+  }
+};
+
+// Class methods.
+// Method to get the joint state information from publisher node.
+void SensorMeasurementData::joint_statesCallback(const sensor_msgs::JointState::ConstPtr& msg)
 {
-  ROS_INFO("I heard: [%s]", msg->data.c_str());
+  ROS_INFO("I heard: [%F]", msg->position[0]);
+
+  position_joint1_ = msg->position[0];
+  position_joint2_ = msg->position[1];
+
+  ee_position_ = eePos(position_joint1_, position_joint2_);
+
+  saveYaml();
 }
 
-int main (int argc, char **argv){
+// Method to write the joint state angles and end effector position to yaml file.
+void SensorMeasurementData::saveYaml()
```

Comment:
I don't know how to open a file, write to a file and closing the file in different functions. :( 
I spent some time researching this today but I'm still not sure since the fstream object would be implicitly deleted at the end of each function.